### PR TITLE
Support query pass-through for BigQuery connector

### PR DESCRIPTION
**Pull request by ebyhr**
## Description

Support query pass-through for BigQuery connector

## Documentation

( ) No documentation is needed.
( ) Sufficient documentation is included in this PR.
( ) Documentation PR is available with #prnumber.
(x) Documentation issue #issuenumber is filed, and can be handled later.

## Release notes

( ) No release notes entries required.
(x) Release notes entries required with the following suggested text:

```markdown
# BigQuery
* Support query pass-through table function. ({issue}`12502`)
```

### CODE REVIEW

**Comment by mosabua** [comment]
This will need docs .. @colebow can you maybe create a PR with the help of @ebyhr for technical input.

**Comment by kasiafi** [comment]
>then this is just plumbing without user facing impact.

@mosabua @colebow This is a concrete Table Function that will be exposed by the BigQuery connector, so it definitely needs to be documented.

**Comment by mosabua** [comment]
Thanks for confirming @kasiafi !

**Comment by mosabua** [comment]
fyi @brianzhan

**Comment by ebyhr** [comment]
Removing `syntax-needs-review` label since JDBC connectors' PR was already merged.

**Review by mosabua — APPROVED** [review]
I only looked at the docs .. they are good to go.

**Comment by ebyhr** [comment]
@hashhar @wendigo Gentle reminder.

**Review by hashhar — APPROVED** [review]
Looks good to me however I would appreciate if @kasiafi skims as well.

The tests look good candidates for BCT now that we have query PTF in both JDBC and non-JDBC connectors - I'll see if I can come up with something.

File: plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryClient.java [hunk 1]
```diff
@@ -224,6 +229,27 @@ public TableResult query(String sql, boolean useQueryResultsCache, CreateDisposi
...
+        try {
+            statistics = bigQuery.create(jobInfo).getStatistics();
+        }
+        catch (BigQueryException e) {
+            throw new TrinoException(BIGQUERY_INVALID_STATEMENT, "Failed to get schema for query: " + sql, e);
+        }
+
+        QueryStatistics queryStatistics = (QueryStatistics) statistics;
+        if (!queryStatistics.getStatementType().equals(SELECT)) {
+            throw new TrinoException(BIGQUERY_INVALID_STATEMENT, "Unsupported statement type: " + queryStatistics.getStatementType());
```
[the remark this answers is not in the document]
Comment (mosabua) [reply plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryClient.java:247]: Can you make sure we got this covered @colebow .. thx

File: plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryMetadata.java [hunk 2]
```diff
@@ -468,6 +493,24 @@ public Optional<ConstraintApplicationResult<ConnectorTableHandle>> applyFilter(
         return Optional.of(new ConstraintApplicationResult<>(updatedHandle, constraint.getSummary(), false));
     }
 
+    @Override
+    public Optional<TableFunctionApplicationResult<ConnectorTableHandle>> applyTableFunction(ConnectorSession session, ConnectorTableFunctionHandle handle)
```
Comment (ebyhr) [inline plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryMetadata.java:497]: @martint @kasiafi Could you please review around `applyTableFunction()` in this PR when you have time?

Comment (ebyhr) [reply plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryMetadata.java:497]: @martint @kasiafi Gentle reminder.

File: docs/src/main/sphinx/connector/bigquery.rst [hunk 3]
```diff
@@ -218,6 +218,39 @@ the following features:
...
+
+The connector provides specific :doc:`table functions </functions/table>` to
+access BigQuery.
+
+.. _bigquery-query-function:
+
+``query(varchar) -> table``
+^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The ``query`` function allows you to query the underlying database directly. It
```
Comment (wendigo) [inline docs/src/main/sphinx/connector/bigquery.rst:232]: The ``query`` function allows you to query the underlying BigQuery directly.

File: docs/src/main/sphinx/connector/bigquery.rst [hunk 4]
```diff
@@ -218,6 +218,39 @@ the following features:
...
+requires syntax native to BigQuery, because the full query is pushed down and
+processed in BigQuery. This can be useful for accessing native features which are
```
Comment (wendigo) [inline docs/src/main/sphinx/connector/bigquery.rst:234]: processed by

File: plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryMetadata.java [hunk 5]
```diff
@@ -221,7 +222,12 @@ public ConnectorTableHandle getTableHandle(ConnectorSession session, SchemaTable
             return null;
         }
 
-        return new BigQueryTableHandle(schemaTableName, new RemoteTableName(tableInfo.get().getTableId()), tableInfo.get());
+        return new BigQueryTableHandle(new BigQueryNamedRelationHandle(
+                schemaTableName,
+                new RemoteTableName(tableInfo.get().getTableId()),
+                tableInfo.get().getDefinition().getType().toString(),
+                getPartitionType(tableInfo.get().getDefinition()),
+                Optional.ofNullable(tableInfo.get().getDescription())));
```
Comment (wendigo) [inline plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryMetadata.java:235]: This commit adds more then just relation handle i.e. it adds support for table comments.

File: plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryClient.java [hunk 6]
```diff
@@ -224,6 +229,27 @@ public TableResult query(String sql, boolean useQueryResultsCache, CreateDisposi
...
+        }
+
+        return requireNonNull(queryStatistics.getSchema(), "Result schema not provided for query");
```
Comment (wendigo) [inline plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryClient.java:250]: Cannot determine schema for query

File: plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/ptf/Query.java [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+                }
+                columnsBuilder.add(toColumnHandle(field));
+            }
+            List<BigQueryColumnHandle> columns = columnsBuilder.build();
+
+            Descriptor returnedType = new Descriptor(columnsBuilder.build().stream()
+                    .map(column -> new Field(column.getName(), Optional.of(column.getTrinoType())))
+                    .collect(toList()));
+
+            QueryHandle handle = new QueryHandle(tableHandle.withProjectedColumns(columns.stream().map(c -> (ColumnHandle) c).collect(toList())));
```
Comment (wendigo) [inline plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/ptf/Query.java:119]: s/c/column

**Review by wendigo — COMMENTED** [review]
Overall design looks good to me.

File: plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryMetadata.java (hunk 5 above)
Comment (ebyhr) [reply plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryMetadata.java:235]: Table comment is already supported in BigQuery connector https://github.com/trinodb/trino/blob/35f68312ed7dc065245d4d55d9420e10d74069f8/plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryTableHandle.java#L54

**Comment by ebyhr** [comment]
Let me merge this PR if there's no additional comments until tomorrow.

**Pull request merged**